raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2268,6 +2269,25 @@ private Long append(int epoch, List<T> records, boolean isAtomic) {
         );
     }
 
+    private void validateSnapshotId(OffsetAndEpoch snapshotId) {
+        Optional<LogOffsetMetadata> highWatermarkOpt = quorum().highWatermark();
+        if (!highWatermarkOpt.isPresent() || highWatermarkOpt.get().offset < snapshotId.offset) {
+            throw new KafkaException("Trying to creating snapshot with invalid snapshotId: " + snapshotId + " whose offset is larger than the high-watermark: " +
+                    highWatermarkOpt + ". This may necessarily mean a bug in the caller, since the there should be a minimum " +
+                    "size of records between the latest snapshot and the high-watermark when creating snapshot");
+        }
+        int quorumEpoch = quorum().epoch();
+        if (snapshotId.epoch > quorumEpoch) {
+            throw new KafkaException("Trying to creating snapshot with invalid snapshotId: " + snapshotId + " whose epoch is" +
+                    " larger than the quorum epoch: " + quorumEpoch);
+        }
+        OffsetAndEpoch endOffsetAndEpoch = log.endOffsetForEpoch(snapshotId.epoch);
+        if (endOffsetAndEpoch.epoch != snapshotId.epoch || endOffsetAndEpoch.offset < snapshotId.offset) {
+            throw new KafkaException("Trying to creating snapshot with invalid snapshotId: " + snapshotId +
```

Comment:
Let's use `InvalidArgumentException` instead of `KafkaException` for all of the exceptions thrown in this method.

Reply:
Will fix, but I assume you want `IllegalArgumentException` instead of `InvalidArgumentException`?

Reply:
Yes. `IllegalArgumentException`.